src/v/cluster/members_backend.cc

```diff
@@ -43,6 +44,183 @@
 #include <vector>
 
 namespace cluster {
+namespace {
+struct node_replicas {
+    size_t allocated_replicas;
+    size_t max_capacity;
+};
+struct unevenness_error_info {
+    double e;
+    double e_step;
+};
+using node_replicas_map_t = absl::node_hash_map<model::node_id, node_replicas>;
+static absl::node_hash_map<model::node_id, node_replicas>
+calculate_replicas_per_node(
+  const partition_allocator& allocator, partition_allocation_domain domain) {
+    node_replicas_map_t ret;
+    ret.reserve(allocator.state().allocation_nodes().size());
+
+    for (const auto& [id, n] : allocator.state().allocation_nodes()) {
+        if (!n->is_active()) {
+            continue;
+        }
+        auto [it, _] = ret.try_emplace(
+          id,
+          node_replicas{
+            .allocated_replicas = 0,
+            .max_capacity = n->domain_partition_capacity(domain),
+          });
+
+        const auto domain_allocated = n->domain_allocated_partitions(domain);
+        it->second.allocated_replicas += domain_allocated;
+    }
+    return ret;
+}
+
+static size_t
+calculate_total_replicas(const node_replicas_map_t& node_replicas) {
+    size_t total_replicas = 0;
+    for (auto& [_, replicas] : node_replicas) {
+        total_replicas += replicas.allocated_replicas;
+    }
+    return total_replicas;
+}
+
+void reassign_replicas(
+  partition_allocator& allocator,
+  partition_assignment& current_assignment,
+  members_backend::partition_reallocation& reallocation) {
+    // remove nodes that are going to be reassigned from current assignment.
+    std::erase_if(
+      current_assignment.replicas,
```

Comment:
nit: current_assignment can be passed by value I think, this way there is no impression that this function is modifying some external variable before the actual reallocation takes place